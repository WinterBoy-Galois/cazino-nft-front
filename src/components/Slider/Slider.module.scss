@import '../../styles/shared.scss';

.slider {
  position: relative;
  height: 100%;
  user-select: none;

  &__rail {
    height: 100%;
    width: 100%;
    cursor: pointer;

    &--green {
      background-color: $green;
    }

    &--red {
      background-color: $red;
    }

    &--disabled {
      cursor: default;
    }
  }

  &__wrapper {
    height: 100%;
    width: 100%;
  }
}
