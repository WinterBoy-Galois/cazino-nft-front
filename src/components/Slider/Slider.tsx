import React, { useEffect, useState } from 'react';
import styles from './Slider.module.scss';
import { Slider as ReactSlider, Rail, Handles, Tracks } from 'react-compound-slider';
import Handle from './components/Handle';
import Track from './components/Track';
import useSound from 'use-sound';

import { useStateValue } from '../../state';
import { dice_slider_v1 } from '../../components/App/App';

interface IProps {
  value?: number;
  disabled?: boolean;
  onUpdate?: (value: number) => void;
  onChange?: (value: number) => void;
  switchColors?: boolean;
  valueRange?: number[];
  maxValue?: number;
  minValue?: number;
}

const Slider: React.FC<IProps> = ({
  value = 50,
  disabled = false,
  onUpdate,
  onChange,
  switchColors = false,
  valueRange = [0, 100],
  maxValue = 100,
  minValue = 0,
}: IProps) => {
  const [steps, setSteps] = useState(0.01);

  const [playDiceSlider] = useSound(dice_slider_v1.default);
  const [
    {
      sidebar: { isSound },
    },
  ] = useStateValue();

  useEffect(() => {
    if (value % 1 != 0) {
      setSteps(0.01);
    }
    if (isSound) {
      (async () => {
        await playDiceSlider();
      })();
    }
  }, [value]);

  return (
    <div className={styles.slider__wrapper}>
      <ReactSlider
        mode={(_, next) => {
          if (maxValue && minValue) {
            const nextVal = next[0]?.val;
            const nextKey = next[0]?.key;

            if (nextVal > maxValue) {
              return [
                {
                  key: nextKey,
                  val: maxValue,
                },
              ];
            }

            if (nextVal < minValue) {
              return [
                {
                  key: nextKey,
                  val: minValue,
                },
              ];
            }
          }

          return next;
        }}
        vertical
        reversed
        step={steps}
        domain={valueRange}
        onUpdate={(u: any) =>
          onUpdate && u.length > 0 && u[0] !== +value.toFixed(2) && u[0] !== 0 && onUpdate(u[0])
        }
        onChange={(c: any) =>
          onChange && c.length > 0 && c[0] !== +value.toFixed(2) && c[0] !== 0 && onChange(c[0])
        }
        values={[value]}
        className={styles.slider}
        disabled={disabled}
      >
        <Rail>
          {() => (
            <div
              className={`${styles.slider__rail} ${
                switchColors ? styles['slider__rail--red'] : styles['slider__rail--green']
              } ${disabled ? styles['slider__rail--disabled'] : ''}`}
            />
          )}
        </Rail>

        <Handles>
          {({ handles, getHandleProps }) => (
            <div className={styles['slider-handles']}>
              {handles.map(handle => (
                <Handle
                  key={handle.id}
                  handle={handle}
                  domain={valueRange}
                  getHandleProps={getHandleProps}
                  disabled={disabled}
                  onTouchHandle={() => {
                    if (value < minValue) {
                      onUpdate && onUpdate(minValue);
                    }

                    if (value > maxValue) {
                      onUpdate && onUpdate(maxValue);
                    }

                    setSteps(1);
                  }}
                />
              ))}
            </div>
          )}
        </Handles>

        <Tracks right={false}>
          {({ tracks, getTrackProps }) => (
            <div className={styles['slider-tracks']}>
              {tracks.map(({ id, source, target }) => (
                <Track
                  key={id}
                  source={source}
                  target={target}
                  getTrackProps={getTrackProps}
                  switchColors={switchColors}
                  disabled={disabled}
                />
              ))}
            </div>
          )}
        </Tracks>
      </ReactSlider>
    </div>
  );
};

export default Slider;
