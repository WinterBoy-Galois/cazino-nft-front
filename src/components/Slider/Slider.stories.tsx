import React from 'react';
import { storiesOf } from '@storybook/react';
import Slider from '.';
import { withKnobs, boolean } from '@storybook/addon-knobs';

storiesOf('Components/Slider', module)
  .addDecorator(withKnobs)
  .addDecorator(storyFn => (
    <div
      style={{
        display: 'flex',
        justifyContent: 'center',
        height: '500px',
        width: '70px',
        margin: '0 auto',
        paddingTop: '30px',
      }}
    >
      {storyFn()}
    </div>
  ))
  .add('default', () => (
    <Slider disabled={boolean('Disabled', false)} switchColors={boolean('Switch colors', false)} />
  ));
