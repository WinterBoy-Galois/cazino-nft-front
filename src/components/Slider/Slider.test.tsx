import React from 'react';
import { render } from '@testing-library/react';
import Slider from '.';

test('renders', () => {
  const { container } = render(<Slider />);
  expect(container).toMatchSnapshot();
});
