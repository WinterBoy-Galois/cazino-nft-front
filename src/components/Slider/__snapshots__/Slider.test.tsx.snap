// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div>
  <div
    class="slider__wrapper"
  >
    <div
      class="slider"
    >
      <div
        class="slider__rail slider__rail--green "
      />
      <div
        class="slider-handles"
      >
        <svg
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="50"
          class="handle "
          height="41.562"
          role="slider"
          style="top: 50%; user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
          width="88.143"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M26.451 41.547h-.064V41.5L0 20.774 26.387.053V0H80.142a8 8 0 018 8v25.547a8 8 0 01-8 8H26.471v.015z"
            fill="#2a434f"
          />
          <text
            class="handle__value"
            x="24%"
            y="70%"
          >
            50.00
          </text>
        </svg>
      </div>
      <div
        class="slider-tracks"
      >
        <div
          class="track track--red "
          style="top: 0%; height: 50%;"
        />
      </div>
    </div>
  </div>
</div>
`;
